[APM] Alert counts in Service groups

ogupte opened the pull request:
Closes #143613

Closes https://github.com/elastic/kibana/issues/144420

![Screen Shot 2022-11-03 at 7 32 39 PM](https://user-images.githubusercontent.com/1967266/199854883-9b4b5028-c2c6-46ca-93ce-cce37e31a213.png)
![apm-service-groups-alert-count](https://user-images.githubusercontent.com/1967266/199854840-70c17d59-5594-46c4-8fcb-d3e39e149d27.gif)
![Screen Shot 2022-11-03 at 7 34 41 PM](https://user-images.githubusercontent.com/1967266/199854863-149c638a-e978-41a7-bc3d-ccf1ccc7c53b.png)
![Screen Shot 2022-11-03 at 7 32 21 PM](https://user-images.githubusercontent.com/1967266/199854876-b49249b8-bfa7-4106-a9e8-632c794cffde.png)

ogupte committed:
[APM] Alert counts in Service groups (#143613)

ogupte committed:
cleanup

ogupte committed:
- moves dynamic: true mapping to specific labels field rather than top-level mapping
- adds click thru for alerts count badge to observability alerts paged filtered by service group kuery

ogupte committed:
combined service group services and alerts count into 1 api call

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

elasticmachine commented:
Pinging @elastic/apm-ui (Team:APM)

ogupte committed:
fixes linting errors

yngrdyn commented on x-pack/plugins/apm/server/plugin.ts:
```diff
@@ -132,8 +134,18 @@ export class APMPlugin
+              [SERVICE_LANGUAGE_NAME]: {
+                type: 'keyword',
+              },
+              labels: {
```
Should we include this is `elasticsearch_fieldnames.ts`?
I see we have there `LABEL_NAME = 'labels.name'`

sqren replied:
Yes, @MiriamAparicio, will you do that?

MiriamAparicio replied:
@sqren but this is just `labels`, not `labels.name`, or am I missing something?

yngrdyn commented on x-pack/plugins/apm/public/components/app/service_groups/service_group_save/create_button.tsx:
```diff
@@ -7,42 +7,21 @@
 import { EuiButton } from '@elastic/eui';
 import { i18n } from '@kbn/i18n';
 import React from 'react';
-// import { ServiceGroupsTour } from '../service_groups_tour';
```
Can we get rid of this?

pmuellr reviewed (approved):
LGTM

dgieselaar commented on x-pack/plugins/rule_registry/common/mapping_from_field_map.ts:
```diff
@@ -11,7 +11,7 @@ import { FieldMap } from './field_map/types';
 export function mappingFromFieldMap(
   fieldMap: FieldMap,
-  dynamic: 'strict' | boolean
+  dynamic: estypes.MappingDynamicMapping
```
do you still need this change?

ogupte replied:
reverted

dgieselaar reviewed (commented):
@ogupte can you add API tests?

ymao1 commented on x-pack/plugins/rule_registry/common/mapping_from_field_map.ts:
```diff
@@ -11,7 +11,7 @@ import { FieldMap } from './field_map/types';
 export function mappingFromFieldMap(
   fieldMap: FieldMap,
-  dynamic: 'strict' | boolean
+  dynamic: estypes.MappingDynamicMapping
```
We would prefer not to turn on dynamic mapping for these alerts as data indices so that we can control which indices are getting mapped. It looks like you're using `dynamic: true` in order to avoid explicitly mapping the `labels` object? Is the `labels` object ECS compliant? If so, can I request we explicitly add the mappings for that object for the APM rule registry indices?

ogupte replied:
the `labels` space is reserved for custom fields in ECS: https://www.elastic.co/guide/en/ecs/8.6/ecs-base.html#field-labels

ymao1 replied:
Gotcha. I did not realize. Will the data underneath labels need to be searchable/aggregatable? Is it possible to keep this object defined but unmapped?

ogupte replied:
For this feature, we need to support filtering by `label.*` that users use frequently in the form of KQL.

ymao1 replied:
That's fine! The default mapping for the `labels` field in ECS is `type: object` [reference](https://github.com/elastic/ecs/blob/main/generated/elasticsearch/legacy/template.json) I did a quick test and it looks like just using the `object` mapping allows sub-properties to be indexed and searchable.

cauemarcondes commented on x-pack/plugins/apm/public/components/app/service_groups/service_groups_list/service_group_card.tsx:
```diff
@@ -58,15 +83,15 @@ export function ServiceGroupsCard({
             <EuiText size="s">
-              {servicesCount === undefined ? (
+              {serviceGroupCounts === undefined ? (
                 <>&nbsp;</>
```
Is it really needed? Can we just render the translation if `serviceCount`is defined?

ogupte replied:
This was in place for 8.5 in order to avoid a jump in height after service counts were loaded. But we may want to re-think the loading state for these cards of service groups.

cauemarcondes commented on x-pack/plugins/apm/public/components/app/service_groups/service_groups_list/service_groups_list.tsx:
```diff
@@ -29,7 +30,7 @@ export function ServiceGroupsListItems({ items, servicesCounts }: Props) {
       {items.map((item) => (
         <ServiceGroupsCard
```
You probably need a `key` property in here, don't you? 

```suggestion
        <ServiceGroupsCard key={item.id}
```

cauemarcondes commented on x-pack/plugins/apm/server/routes/service_groups/get_services_counts.ts:
```diff
@@ -67,10 +67,14 @@ export async function getServicesCounts({
     response?.aggregations?.service_groups.buckets ?? {};
-  return Object.keys(buckets).reduce((acc, key) => {
+
+  const servicesCount: Record<string, number> = Object.keys(buckets).reduce<
```
Why store the result in a variable if you're not using it, can't you just return it? And type the function return?

cauemarcondes commented on x-pack/plugins/apm/server/routes/service_groups/route.ts:
```diff
@@ -188,12 +154,57 @@ const serviceGroupServicesRoute = createApmServerRoute({
+        context,
+      }),
+    ]);
+    const serviceGroupCounts: ServiceGroupCounts = serviceGroups.reduce(
+      (acc, { id }): ServiceGroupCounts => ({
```
I think you can instead type the reduce.

```suggestion
    const serviceGroupCounts: ServiceGroupCounts = serviceGroups.reduce<ServiceGroupCounts>(
      (acc, { id })  => ({
```

ogupte replied:
changing this results in a linting error:
```
190:7  error  Missing return type on function  @typescript-eslint/explicit-function-return-type
```

cauemarcondes reviewed (commented):
Some suggestions.

ogupte committed:
PR feedback

ymao1 commented on x-pack/plugins/apm/server/plugin.ts:
```diff
@@ -132,8 +134,18 @@ export class APMPlugin
+              },
+              labels: {
+                type: 'object',
+                dynamic: true,
```
dynamic defaults to `true` for the object field type so I don't think this is strictly necessary?

ogupte replied:
When i take away `dynamic: true`, my aggregation fails:
```
GET .alerts-observability.apm.alerts-*/_search
{
  "size": 10,
  "query": {
    "bool": {
      "filter": [
        {
          "term": {
            "kibana.alert.rule.producer": "apm"
          }
        },
        {
          "term": {
            "kibana.alert.status": "active"
          }
        }
      ]
    }
  },
  "aggs": {
    "service_groups": {
      "filters": {
        "filters": {
          "2641b250-5aba-11ed-a895-07dc9e2f77a8": {
            "bool": {
              "should": [
                {
                  "match": {
                    "labels.country_code": "US"
                  }
                }
              ],
              "minimum_should_match": 1
            }
          }
        }
      },
      "aggs": {
        "alerts_count": {
          "cardinality": {
            "field": "kibana.alert.uuid"
          }
        }
      }
    }
  }
}
```
Returns:
```
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 6,
      "relation": "eq"
    },
    "max_score": 0,
    "hits": [
      {
        "_index": ".internal.alerts-observability.apm.alerts-default-000001",
        "_id": "ff1ef1f2-8ad7-4949-b610-c11bd5dfe2c7",
        "_score": 0,
        "_source": {
          "kibana.alert.rule.category": "Error count threshold",
          "kibana.alert.rule.consumer": "alerts",
          "kibana.alert.rule.execution.uuid": "46ecf66e-6c55-4806-89db-ca67ae659242",
          "kibana.alert.rule.name": "> 25 errors last 1 min",
          "kibana.alert.rule.producer": "apm",
          "kibana.alert.rule.rule_type_id": "apm.error_rate",
          "kibana.alert.rule.uuid": "090e77d0-50b1-11ed-8299-8da1b02a8d3b",
          "kibana.space_ids": ["default"],
          "kibana.alert.rule.tags": [],
          "@timestamp": "2022-11-06T16:07:54.794Z",
          "service.name": "elastic-co-frontend",
          "service.environment": "https://www.elastic.co",
          "processor.event": "error",
          "kibana.alert.evaluation.value": 29,
          "kibana.alert.evaluation.threshold": 25,
          "kibana.alert.reason": "Error count is 29 in the last 1 min for elastic-co-frontend. Alert when > 25.",
          "agent.name": "rum-js",
          "service.language.name": "javascript",
          "labels.in_eu": "false",
          "labels.country_code": "US",
          "labels.city": "council bluffs",
          "labels.ip": "35.224.28.11",
          "labels.lang": "en-US",
          "kibana.alert.duration.us": 312414000,
          "kibana.alert.time_range": {
            "gte": "2022-11-06T16:02:42.380Z"
          },
          "kibana.alert.instance.id": "apm.error_rate_elastic-co-frontend_https://www.elastic.co",
          "kibana.alert.start": "2022-11-06T16:02:42.380Z",
          "kibana.alert.uuid": "ff1ef1f2-8ad7-4949-b610-c11bd5dfe2c7",
          "kibana.alert.status": "active",
          "kibana.alert.workflow_status": "open",
          "event.kind": "signal",
          "event.action": "active",
          "tags": [],
          "kibana.version": "8.6.0"
        }
      },
      ...
    ]
  },
  "aggregations": {
    "service_groups": {
      "buckets": {
        "2641b250-5aba-11ed-a895-07dc9e2f77a8": {
          "doc_count": 0,
          "alerts_count": {
            "value": 0
          }
        }
      }
    }
  }
}
```
Notice how the document has field `"labels.country_code": "US"`, but the filters agg has doc_count: 0.

The mapping shows ` "labels": { "type": "object" },`:
```
{
  ".internal.alerts-observability.apm.alerts-default-000001": {
    "mappings": {
      "dynamic": "false",
      "_meta": {
        "namespace": "default",
        "kibana": { "version": "8.6.0" }
      },
      "properties": {
        "@timestamp": { "type": "date" },
        "agent": { "properties": {...} },
        "ecs": { "properties": {...} },
        "event": { "properties": {...} },
        "kibana": { "properties": {...} },
        "labels": { "type": "object" },
        "processor": { "properties": {...} },
        "service": {...},
        "tags": { "type": "keyword" },
        "transaction": { "properties": {...} }
      }
    }
  }
}
```
But keeping the `dynamic: true` here, allows the labels.country_code field to be accessible to the filters agg as expected.

dgieselaar replied:
@ymao1 `dynamic` is inherited from its parents, and only if it is entirely omitted from the mapping, including at the top level, the value defaults to `true`.

ymao1 replied:
Gotcha, thanks! I'm fine with setting this to true if the top level setting is changed back to `strict`

ymao1 commented on x-pack/plugins/apm/server/plugin.ts:
```diff
@@ -132,8 +134,18 @@ export class APMPlugin
+              },
             },
-            'strict'
+            true
```
Can we change this back to `strict`?

dgieselaar commented on x-pack/plugins/apm/server/routes/service_groups/get_service_group_alerts.ts:
```diff
@@ -0,0 +1,92 @@
+      },
+    },
+  };
+  const result = await esClient.asCurrentUser.search(params);
```
How does this work? These are supposed to be system indices that are not accessible to the end user. Have you tried this with non-superusers?

ogupte replied:
Right now if the user doesn't have access to query the internal index, it just omits the alert counts from the response. I can look into the rule Data client, if it already handles this, or we could use the internal user as a last resort. 
I also discovered a related issue that Service groups breaks down totally if the user doesn't have kibana read privileges for the 'Saved objects management' feature.

dgieselaar replied:
I don't think this is how those indices should be accessed. They should be accessed using the internal user, and there should be an evaluation before that whether the user has access to the alerts feature. Non-superusers will not have access to these indices (and perhaps even superusers won't, I remember there was a recent change wrt system indices). I don't think omitting the alert count in many cases is a good workaround. Can you reach out to the alerting teams to figure out how we should handle it if we want to read from the alerts indices?

dgieselaar commented on x-pack/plugins/apm/server/routes/service_groups/get_service_group_alerts.ts:
```diff
@@ -0,0 +1,92 @@
+      };
+    }, {});
+
+  const esClient = (await context.core).elasticsearch.client;
```
Doesn't the ruleDataClient expose a `getReader()` function that you can use, with the indices predefined?

dgieselaar replied:
Is there any prior art here as to how other teams are doing this?

ogupte replied:
the `alertsClient.find` supports arbitrary aggregations, and handles RBAC and active space filter for us.

ogupte committed:
addresses feedback:
- checks user has privileges for alerts index before querying with internal user
- filters alert by active space id

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

ogupte committed:
querying with the alertsClient.find rather than esClient directly

sqren commented on x-pack/plugins/apm/server/routes/service_groups/get_apm_alerts_client.ts:
```diff
@@ -0,0 +1,34 @@
+  const alertsClient = await ruleRegistryPluginStart.getRacClientWithRequest(
+    request
+  );
+  const apmAlertsIndices = (await alertsClient.getAuthorizedAlertsIndices([
+    'apm',
+  ])) ?? ['.alerts*'];
```
When would `alertsClient.getAuthorizedAlertsIndices` return null?

sqren replied:
nit
```suggestion
  const apmAlertsIndices = (await alertsClient.getAuthorizedAlertsIndices([
    'apm',
  ]));
  
  const index = apmAlertsIndices?.[0] ?? '.alerts*';
```

ogupte replied:
>When would alertsClient.getAuthorizedAlertsIndices return null?

`undefined` is on of the return types, and it looks like typescript resolves it as undefined when it `throw`s

sqren commented on x-pack/plugins/apm/server/routes/service_groups/get_apm_alerts_client.ts:
```diff
@@ -0,0 +1,34 @@
+
+  return {
+    search(searchParams: ApmAlertsClientSearchParams) {
+      return alertsClient.find({ index: apmAlertsIndices[0], ...searchParams });
```
Do we always only want the first one? When are there multiple?

ogupte replied:
maybe we should `].join(',')` them

sqren commented on x-pack/plugins/apm/server/routes/service_groups/get_service_group_alerts.ts:
```diff
@@ -0,0 +1,91 @@
+    query: {
+      bool: {
+        filter: [
+          { term: { 'kibana.alert.rule.producer': 'apm' } },
+          { term: { 'kibana.alert.status': 'active' } },
```
WDYT about storing the field names as constants, similar to `elasticsearch_fieldnames.ts`?

fkanout reviewed (approved):
Actionable Observability changes LGTM

ogupte committed:
PR feedback

ogupte committed:
joins multiple alerts indices

ymao1 commented on x-pack/plugins/apm/server/routes/service_groups/get_service_group_alerts.ts:
```diff
@@ -0,0 +1,95 @@
+import {
+  KIBANA_ALERT_RULE_PRODUCER,
+  KIBANA_ALERT_STATUS,
+} from '../../../common/rules/apm_rule_types';
```
Could we just use `ALERT_RULE_PRODUCER` and `ALERT_STATUS` from `@kbn/rule-data-utils`

ymao1 commented on x-pack/plugins/rule_registry/common/field_map/types.ts:
```diff
@@ -12,5 +12,6 @@ export interface FieldMap {
     array?: boolean;
     path?: string;
     scaling_factor?: number;
+    dynamic?: 'strict' | boolean;
```
this change is no longer needed right?

ogupte replied:
It's still needed to set `dynamic: true` on the field mapping for the reason [stated above](https://github.com/elastic/kibana/pull/144484#discussion_r1014857283)

ogupte committed:
adds stubs to for API tests

ogupte committed:
PR feedback and adds API tests

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

ogupte committed:
- improves API tests for service group counts
- updates save service group API to return the saved service group

MiriamAparicio committed:
Merge branch 'main' into apm-143613-service-groups-alert-count

dgieselaar commented on x-pack/plugins/apm/server/lib/helpers/get_apm_event_client.ts:
```diff
@@ -32,7 +32,7 @@ export async function getApmEventClient({
 
     return new APMEventClient({
-      esClient: coreContext.elasticsearch.client.asCurrentUser,
+      esClient: coreContext.elasticsearch.client.asInternalUser,
```
we definitely do not want to do this. good catch!

MiriamAparicio replied:
I'll fix this, good catch!

ogupte replied:
wow whoops...i did this to test something during development and forgot about it...great catch!

MiriamAparicio committed:
fix apmEventClient user

dgieselaar commented on x-pack/plugins/apm/server/routes/service_groups/get_service_group_alerts.ts:
```diff
@@ -0,0 +1,92 @@
+      },
+    },
+    aggs: {
+      service_groups: {
+        filters: {
+          filters: serviceGroupsKueryMap,
+        },
```
I don't think this needs to be an aggregation? why not add it to `query.bool.filter`?

dgieselaar replied:
ahh, it's a map and a `filters` agg, so it returns buckets. never mind, that's good!

sqren commented on x-pack/plugins/apm/server/lib/helpers/get_apm_event_client.ts:
```diff
@@ -32,7 +32,7 @@ export async function getApmEventClient({
 
     return new APMEventClient({
-      esClient: coreContext.elasticsearch.client.asInternalUser,
+      esClient: coreContext.elasticsearch.client.asCurrentUser,
```
As mentioned on Slack, we should have an api test that verifies that only users with access to the apm indices can view APM data.

dgieselaar replied:
The internal user has read access to the APM indices

dgieselaar replied:
We can add a test however that runs with a user that has no access to APM data?

sqren replied:
> The internal user has read access to the APM indices

Exactly. So if we add a test with a user that _does not_ have access to APM indices, they should not be able to retrieve APM data.

sqren replied:
https://github.com/elastic/kibana/issues/145516

dgieselaar commented on x-pack/plugins/apm/server/routes/service_groups/get_apm_alerts_client.ts:
```diff
@@ -0,0 +1,37 @@
+  return {
+    search(searchParams: ApmAlertsClientSearchParams) {
+      return alertsClient.find({
+        index: apmAlertsIndices?.join(',') ?? '.alerts*',
```
Do alerts end up in the .alerts* indices or .internal.alerts*?

MiriamAparicio replied:
I'm not sure, what's the difference?

sqren replied:
@dgieselaar I had the same concern: 
https://github.com/elastic/kibana/pull/144484#discussion_r1018977135

@MiriamAparicio It feels problematic that we specify the default index (`.alerts*`) for another teams code here. We should use a constant like `DEFAULT_ALERTS_AS_DATA_INDEX` provided by the rac team.
This ensure that if the underlying index is changed/migrated it will still work for us.

MiriamAparicio replied:
Should then ask rac team to add this constant?

sqren replied:
yes, ask the rac team what they recommend for querying alerts.

MiriamAparicio committed:
fix conflicts

MiriamAparicio committed:
fix conflicts

MiriamAparicio committed:
abort if alert indices are not found

sqren commented on x-pack/plugins/apm/server/routes/service_groups/get_apm_alerts_client.ts:
```diff
@@ -0,0 +1,42 @@
+    'apm',
+  ]);
+
+  if (!apmAlertsIndices || isEmpty(apmAlertsIndices)) {
```
Is the first check needed?

```suggestion
  if (isEmpty(apmAlertsIndices)) {
```

MiriamAparicio replied:
Yes, `isEmpty` is not checking if `apmAlertIndices` is undefined, and the type is complaining in line 37

sqren commented on x-pack/plugins/apm/server/routes/service_groups/get_apm_alerts_client.ts:
```diff
@@ -0,0 +1,42 @@
+  ]);
+
+  if (!apmAlertsIndices || isEmpty(apmAlertsIndices)) {
+    throw Error('No alert indices exist for "apm". Please fix it');
```
Let's remove the last bit :)

```suggestion
    throw Error('No alert indices exist for "apm".');
```

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/save_service_group.spec.ts:
```diff
@@ -65,12 +65,16 @@ export default function ApiTest({ getService }: FtrProviderContext) {
         kuery: 'service.name: synth*',
-      });
+      };
+      const response = await callApi(serviceGroup);
```
Would be nice if this was more clear:

```suggestion
      const response = await createServiceGroupApi(serviceGroup);
```

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/save_service_group.spec.ts:
```diff
@@ -65,12 +65,16 @@ export default function ApiTest({ getService }: FtrProviderContext) {
+      expect(response.body).to.have.property('id');
+      expect(response.body).to.have.property('groupName', serviceGroup.groupName);
+      expect(response.body).to.have.property('kuery', serviceGroup.kuery);
+      expect(response.body).to.have.property('updatedAt');
```
Shouldn't this test also retrieve the service group to make sure it was created?

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/service_group_count.spec.ts:
```diff
@@ -0,0 +1,206 @@
+  const start = Date.now() - 24 * 60 * 60 * 1000;
+  const end = Date.now();
+
+  async function callApi() {
```
```suggestion
  async function getServiceGroupCountsApi() {
```

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/service_group_count.spec.ts:
```diff
@@ -0,0 +1,206 @@
+    type: string;
+  }>;
+
+  async function deleteServiceGroups() {
```
```suggestion
  async function deleteAllServiceGroups() {
```

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/service_group_count.spec.ts:
```diff
@@ -0,0 +1,206 @@
+  }>;
+
+  async function deleteServiceGroups() {
+    const response = await supertest
+      .get('/api/saved_objects/_find?type=apm-service-group')
+      .set('kbn-xsrf', 'true');
+    const savedObjects: SavedObjectsFindResults = response.body.saved_objects;
+    const bulkDeleteBody = savedObjects.map(({ id, type }) => ({ id, type }));
+    return supertest
+      .post(`/api/saved_objects/_bulk_delete?force=true`)
+      .set('kbn-xsrf', 'true')
+      .send(bulkDeleteBody);
```
Unless it is much slower, we should use our own APIs. 

To retrieve all service groups:
```
GET /internal/apm/service-groups
```

Then, to delete a service group:
```
DELETE /internal/apm/service-group
```

MiriamAparicio replied:
Yes, makes sense, I used our own for the e2e tests

MiriamAparicio replied:
the delete is faster and works better to call the saved_objects

sqren replied:
Yes, I saw :)

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/service_group_count.spec.ts:
```diff
@@ -0,0 +1,206 @@
+                          spanName: '/_search',
+                          spanType: 'db',
+                          spanSubtype: 'elasticsearch',
+                        })
+                        .destination('elasticsearch')
+                        .duration(100)
+                        .success()
+                        .timestamp(timestamp),
+                      service
+                        .span({
+                          spanName: '/_search',
+                          spanType: 'db',
+                          spanSubtype: 'elasticsearch',
+                        })
+                        .destination('elasticsearch')
+                        .duration(300)
+                        .success()
+                        .timestamp(timestamp)
+                    )
+                    .errors(
+                      service.error({ message: 'error 1', type: 'foo' }).timestamp(timestamp),
+                      service.error({ message: 'error 2', type: 'foo' }).timestamp(timestamp),
+                      service.error({ message: 'error 3', type: 'bar' }).timestamp(timestamp)
+                    )
+                )
+            )
+          ),
+          saveServiceGroup({
+            groupName: 'synthbeans',
+            kuery: 'service.name: synth*',
+          }),
+          saveServiceGroup({
+            groupName: 'opbeans',
+            kuery: 'service.name: opbeans*',
+          }),
+        ]
+      );
+      synthbeansServiceGroupId = synthbeansServiceGroup.id;
+      opbeansServiceGroupId = opbeansServiceGroup.id;
+    });
```
We normally extract this outside the test to reduce noise. Can we do the same here? Perhaps create a folder `service_group_count` with both the test and the data generation function

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/service_group_count.spec.ts:
```diff
@@ -0,0 +1,206 @@
+    describe('with alerts', () => {
+      let ruleId: string;
+      before(async () => {
+        const { body: createdRule } = await supertest
+          .post(`/api/alerting/rule`)
+          .set('kbn-xsrf', 'true')
+          .send({
+            params: {
+              serviceName: 'synth-go',
+              transactionType: '',
+              windowSize: 99,
+              windowUnit: 'y',
+              threshold: 100,
+              aggregationType: 'avg',
+              environment: 'testing',
+            },
+            consumer: 'apm',
+            schedule: { interval: '1m' },
+            tags: ['apm'],
+            name: 'Latency threshold | synth-go',
+            rule_type_id: ApmRuleType.TransactionDuration,
+            notify_when: 'onActiveAlert',
+            actions: [],
+          });
+        ruleId = createdRule.id;
+        await waitForActiveAlert({ ruleId, esClient, log });
```
Extract this to `createRule` function

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/wait_for_active_alert.ts:
```diff
@@ -0,0 +1,75 @@
+
+  log.debug(`${message}, retrying`);
+
+  await delay(WAIT_FOR_STATUS_INCREMENT);
+  return await waitForActiveAlert({
+    ruleId,
+    waitMillis: waitMillis - WAIT_FOR_STATUS_INCREMENT,
+    esClient,
+    log,
+  });
```
Can you remove this custom retry logic, and instead use `p-retry`. Perhaps with [a factor](https://github.com/tim-kos/node-retry#retrytimeoutsoptions) of 1 (instead of 2).

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/wait_for_active_alert.ts:
```diff
@@ -0,0 +1,75 @@
+      },
+    },
+  };
+  const response = await esClient.search(searchParams);
```
Can we use the APM alerts client (`getApmAlertsClient`)?

MiriamAparicio replied:
No, I don't think it's an option in within the tests

MiriamAparicio committed:
Merge remote-tracking branch 'upstream/main' into apm-143613-service-groups-alert-count

sqren commented on x-pack/plugins/apm/server/routes/service_groups/get_apm_alerts_client.ts:
```diff
@@ -0,0 +1,42 @@
+  return {
+    search(searchParams: ApmAlertsClientSearchParams) {
+      return alertsClient.find({
+        index: apmAlertsIndices.join(','),
+        ...searchParams,
```
Let's change the order of this so it's not possible to accidentally override `index`

```suggestion
        ...searchParams,
        index: apmAlertsIndices.join(','),
```

MiriamAparicio committed:
PR review comments

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/service_group_count/service_group_count.spec.ts:
```diff
@@ -0,0 +1,159 @@
+          });
+        }
+      });
+      return promises;
```
```suggestion
      return Promise.all(promises);
```

sqren commented on x-pack/test/apm_api_integration/tests/service_groups/save_service_group.spec.ts:
```diff
@@ -44,38 +43,47 @@ export default function ApiTest({ getService }: FtrProviderContext) {
+          });
+        }
+      });
+      return promises;
```
```suggestion
      return Promise.all(promises);
```

sqren replied:
Btw can you extract `deleteAllServiceGroups` so we don't duplicate it

MiriamAparicio committed:
Merge remote-tracking branch 'upstream/main' into apm-143613-service-groups-alert-count

MiriamAparicio committed:
fix promise return

MiriamAparicio committed:
Merge remote-tracking branch 'upstream/main' into apm-143613-service-groups-alert-count

MiriamAparicio committed:
small fix

MiriamAparicio committed:
Merge branch 'main' into apm-143613-service-groups-alert-count

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

MiriamAparicio committed:
Merge branch 'main' into apm-143613-service-groups-alert-count

sorenlouv reviewed (approved):
I'm good with merging this as-is so we can unblock https://github.com/elastic/kibana/issues/146701.
Please create an issue with the follow-up items that were left out.
